src/libos/crates/host-socket/src/common/sockopt.rs

```diff
@@ -0,0 +1,280 @@
+use std::convert::TryFrom;
+
+use async_io::impl_ioctl_cmd;
+use libc::{c_int, c_void, socklen_t};
+use num_enum::{IntoPrimitive, TryFromPrimitive};
+cfg_if::cfg_if! {
+    if #[cfg(feature = "sgx")] {
+        use libc::ocall::getsockopt as do_getsockopt;
+        use libc::ocall::setsockopt as do_setsockopt;
+    } else {
+        use libc::getsockopt as do_getsockopt;
+        use libc::setsockopt as do_setsockopt;
+    }
+}
+
+use crate::prelude::*;
+
+const SOL_SOCKET: i32 = 1;
+
+pub fn new_getsockopt_cmd(
+    level: c_int,
+    optname: c_int,
+    optval: *mut c_void,
+    optlen: *mut socklen_t,
+) -> Result<Box<dyn IoctlCmd>> {
+    if level != SOL_SOCKET {
+        return Ok(Box::new(GetSockOptRawCmd::new(GetSockOptRawInput::new(
+            level, optname, optval, optlen,
+        ))));
+    }
+
+    let opt =
+        SockOptName::try_from(optname).map_err(|_| errno!(ENOPROTOOPT, "Not a valid optname"))?;
+    Ok(match opt {
+        SockOptName::SO_ACCEPTCONN => Box::new(SoAcceptConnCmd::new(())),
+        SockOptName::SO_DOMAIN => Box::new(SoDomainCmd::new(())),
+        SockOptName::SO_PEERNAME => Box::new(SoPeerNameCmd::new(())),
+        SockOptName::SO_TYPE => Box::new(SoTypeCmd::new(())),
+        SockOptName::SO_CNX_ADVICE => return_errno!(ENOPROTOOPT, "it's a write-only option"),
+        _ => Box::new(GetSockOptRawCmd::new(GetSockOptRawInput::new(
+            level, optname, optval, optlen,
+        ))),
+    })
+}
+
+pub fn new_setsockopt_cmd(
```

Comment:
If you think more about it, the `new_getsockopt_cmd` and `new_setsockopt_cmd ` methods belongs more to the LibOS rather than the `host-socket` crate since they serve as a C-to-Rust or Linux-to-host_socket adapter.